sqlmesh/core/engine_adapter/bigquery.py

```diff
@@ -87,6 +88,48 @@ def _job_params(self) -> t.Dict[str, t.Any]:
             params["maximum_bytes_billed"] = self._extra_config.get("maximum_bytes_billed")
         return params
 
+    def _df_to_source_queries(
+        self,
+        df: DF,
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]],
+        batch_size: int,
+        target_table: t.Optional[TableName] = None,
+    ) -> t.Tuple[t.List[SourceQuery], t.Optional[t.Dict[str, exp.DataType]]]:
+        def query_factory(
+            df: pd.DataFrame,
+            columns_to_types: t.Dict[str, exp.DataType],
+            temp_bq_table: bigquery.Table,
+            temp_table: exp.Table,
+        ) -> Query:
+            self._db_call(self.client.create_table, table=temp_bq_table, exists_ok=False)
+            result = self.__load_pandas_to_table(temp_bq_table, df, columns_to_types, replace=False)
+            if result.errors:
+                raise SQLMeshError(result.errors)
+            return exp.select(*columns_to_types).from_(temp_table)
+
+        assert isinstance(df, pd.DataFrame)
+        columns_to_types = columns_to_types or columns_to_types_from_df(df)
+        temp_bq_table = self.__get_temp_bq_table(
+            self._get_temp_table(target_table or "pandas"), columns_to_types
+        )
+        temp_table = exp.table_(
+            temp_bq_table.table_id,
+            db=temp_bq_table.dataset_id,
+            catalog=temp_bq_table.project,
+        )
+        return [
+            SourceQuery(
+                query_factory=partial(
```

Comment:
Why do we need a partial here if `columns_to_types`, `temp_bq_table` and `temp_table` are already local here. Can't `query_factory` reference them directly?

Reply:
How would I pass along the kwargs to include when calling the factory?

Reply:
What I'm saying is that if you define `temp_table`, `temp_bq_table` and `columns_to_types` before you define `query_factory` then you can define `query_factory` without any arguments at all since all of them will be in scope. You can just pass it as is: `query_factory=query_factory`.

Reply:
add kwargs to SourceQuery, and have it call the function with those kwargs on use

Reply:
or, what iarslav was saying is saying it's a lambda, just use closures, and reference them directly

Reply:
Sorry I still don't get it. Putting it in a lambda or passing kwargs to SourceQuery are just two other ways to do functionally what is done here but less clear since they are doing a partial but in other ways.